### ISSUE TITLE
Contact header does not follow RFC 3840

### ISSUE THREAD

**Issue by vasilakisfil** [ByteHeathen/libsip#14]
Hi,

First of all kudos for the parser. I mean, I am surprised how closely you are following the 3261 RFC, you have saved me tons of time. Now to the topic:

I am building a simple SIP server and I was trying to integrate with the Blink SIP client, but I couldn't even make it register, no matter what. I had to check what Kamailio responds to figure out what was missing from the SIP response. The following is produced by current libsip:
```
Contact: sip:56132094@192.168.1.223:5066;+sip.instance=<urn:uuid:1e020c2b-46f6-4867-9d11-65547b8967fa>;expires=600
```
and this is what it should produce:
```
Contact: sip:56132094@192.168.1.223:5066;+sip.instance="<urn:uuid:1e020c2b-46f6-4867-9d11-65547b8967fa>";expires=600
```

As you can see, it's just encapsulating the `+sip.instance` param value under double quotes. It's related to [RFC 3840](https://tools.ietf.org/html/rfc3840#section-9) (specifically `+sip.instance` is defined in [RFC 5626](https://tools.ietf.org/html/rfc5626) but follows the RFC 3840).

I haven't gone through the whole RFC3840, but it seems like the change is simple enough: if the param is well known (defined in RFC 3840) or starts with a plus sign (`+`) then the parser must check the RFC 3840 rules. Rules are simple, but specifically for the case that the value is a string, it **must** be enclosed by `"<` and `>"` (at least that's what I get from the ABNF).

This is the initial REGISTER request btw:

```
REGISTER sip:192.168.1.223 SIP/2.0
Via: SIP/2.0/UDP 192.168.1.223:5066;rport;branch=z9hG4bKPj4f07db29-94d9-4c31-8cc4-652dc54d5ec1
Max-Forwards: 70
From: "vasilakisfil" <sip:vasilakisfil@192.168.1.223>;tag=13727392-aeed-4660-8958-445820a1decf
To: "vasilakisfil" <sip:vasilakisfil@192.168.1.223>
Contact: <sip:32695714@192.168.1.223:5066>;+sip.instance="<urn:uuid:1e020c2b-46f6-4867-9d11-65547b8967fa>"
Call-ID: 07caac9f-6803-46b0-86f4-6c24475a6daa
CSeq: 1 REGISTER
Expires: 600
Supported: gruu
User-Agent: Blink 3.2.1 (Linux)
Content-Length:  0
```

Funny thing is that Blink incorporates a SIP logger, but it wouldn't even show the incoming requests without the proper encapsulation..

**Comment by vasilakisfil** [comment]
I am fine sending a PR, as long as you tell me, roughly, how you would like to solve that.

**Comment by KalitaAlexey** [comment]
Are you planning on an opening a PR? I'll fix it tomorrow if you aren't.

**Comment by KalitaAlexey** [comment]
@vasilakisfil Give me a minute to check.

**Comment by vasilakisfil** [comment]
Cool, I ll take a look tomorrow!

**Comment by KalitaAlexey** [comment]
```
#[test]
fn t() {
    let result =parse_named_field_params::<nom::error::VerboseError<&[u8]>>(";+sip.instance=\"<urn:uuid:1e020c2b-46f6-4867-9d11-65547b8967fa>\"".as_bytes());
    panic!("{:?}", result);
}
``` parsed it.

**Comment by KalitaAlexey** [comment]
@vasilakisfil Given [the definition of gen_value](https://tools.ietf.org/html/rfc3261#page-227), I would suggest that you introduce
```rust
enum GenValue {
Token(String),
QuotedString(String)
}
```
and make it so `fmt` surrounds the value with `"`.

**Comment by vasilakisfil** [comment]
Hi, apparently I am too new in Rust to complete this PR. I did some changes in here: https://github.com/ByteHeathen/libsip/compare/master...vasilakisfil:feat/gen-value

I eventually didn't go for the enum (but I use it internally) because, the way I see it at least, with an enum you have 2 constructors:
* From<String> where I guess you need to check whether the string is wrapped or not
* GenValue::Variant(string) where anyone can manually set the internal string

My thought is that having variants, that are publicly accessible is not a good idea, it might lead to unexpected bugs (for instance setting `GenValue::Token("\"wrapped-string\"")` instead of `GenValue::Token("\"wrapped-string\"")`. So I went for a struct that implements only the From<String> and basically acts as a constructor. It uses the enum you suggested internally, but hides the implementation details.

I have adapted the code to use this struct, but it turns out that we need to modify the parser as well. Specifically, [this](https://github.com/ByteHeathen/libsip/compare/master...vasilakisfil:feat/gen-value#diff-ca58d4c9f0fd4fc2743ad0347807eb6eR126) new test is failing, and I think the underlying changes that need to be done are [here](https://github.com/vasilakisfil/libsip/blob/feat/gen-value/src/headers/parse.rs#L439-L449) and [here](https://github.com/vasilakisfil/libsip/blob/feat/gen-value/src/parse.rs#L103-L110). In theory if we fix this (and the architectural decisions I made make sense), it should be mergable.

As I said I am a bit newcomer to rust and I didn't figure out how to fix the parser, let along my doubts for the whole architecture I went for. Let me know if you have any input.

**Comment by KalitaAlexey** [comment]
@vasilakisfil, I'm a little bit busy with my own project. I took a quick look and couldn't figure out yet. I'll take a look later this week.

**Comment by vasilakisfil** [comment]
yeah no problem! It doesn't seem like it's a trivial change anyway..

**Comment by KalitaAlexey** [comment]
First of all, I'm not a fan of `from_header_tag` returning `GenValue`. I understand that it is allowed by RFC, but in real world I haven't seen tags that aren't just token.

**Comment by KalitaAlexey** [comment]
Can you open a PR so I can leave some comments?
Basically, I still want it to be enum:
```
enum GenValue {
  Token(String)
  QuotedString(String),
}
```
While I see the value in `impl From<String>` I think it's hard to implement and I think it won't be needed for some time (you could construct `GenValue::QuotedString` if needed.

**Comment by vasilakisfil** [comment]
Yeah, I can fix and bring back the enum as you suggest, I think that's better and it's the easy part. What I am not sure about is how to handle the parser. At the moment, the parser, whenever it finds a [quoted string, it just unwraps it](https://github.com/vasilakisfil/libsip/blob/feat/gen-value/src/parse.rs#L103-L110) ).

The `parse_quoted_string` method is used by [`parse_gen_value`](https://github.com/vasilakisfil/libsip/blob/feat/gen-value/src/headers/parse.rs#L439-L449) which returns String. Should we introduce the enum there ? Maybe an easy solution could be to always have String, but that string could be wrapped or not wrapped internally ?

**Comment by KalitaAlexey** [comment]
Take a look how headers are parsed, e.g., map_res alt something

**Comment by bytebuddha** [comment]
@vasilakisfil I'm still looking over your changes now to see whats up with the parser.

> Maybe an easy solution could be to always have String, but that string could be wrapped or not wrapped internally ?

Yes i think this could work We could add a field to NamedHeader that signifies whether the param is wrapped or not

**Comment by vasilakisfil** [comment]
@bytebuddha the first branch is probably way too complicated. I started a new branch [here](https://github.com/ByteHeathen/libsip/compare/master...vasilakisfil:feat/quoted-string?expand=1) by first adding some tests, and then trying to fix the parser for starters without adding the enum. Then I ll introduce the enum. But I am not sure if I am touching the correct parts of the parser. Just now, I am thinking adding a specific contact header parser function, will try it now..

**Comment by bytebuddha** [comment]
Adding a contact header parser function is probably the best way, it's how other headers are implemented. Then you just have to remove the simple macro generated parser in the `_parse_header` method and replace it with the custom one

**Comment by vasilakisfil** [comment]
[here](https://github.com/ByteHeathen/libsip/compare/master...vasilakisfil:feat/quoted-string?expand=1) is a prototype that is working (all my new tests are green). Apparently it's a prototype, what I did is:

I modiefied the `parse_named_field_params` to call `parse_generic_param_with_possibly_quoted_value` (instead of `parse_generic_param`) which calls `parse_gen_possibly_quoted_value` (instead of `parse_gen_value`) that calls `parse_quoted_string_as_quoted` (instead of `parse_quoted_string`)

**Comment by vasilakisfil** [comment]
ok I ll try that

**Comment by vasilakisfil** [comment]
@KalitaAlexey @bytebuddha I created a PR here: https://github.com/ByteHeathen/libsip/pull/16

**Comment by vasilakisfil** [comment]
closed by #16
